### PR TITLE
DRAFT: Multipath support

### PR DESCRIPTION
**Pull request by djosephsen**
An initial take on RT_MULTIPATH message support.

### CODE REVIEW

**Comment by jsimonetti** [comment]
Awesome! Looking forward to review this when ready!

File: route.go [hunk 1]
```diff
@@ -329,3 +335,101 @@ func (rm *RouteMetrics) encode(ae *netlink.AttributeEncoder) error {
...
+			return nil
+		}
+
+		// Copy over the struct portion
+		nh := NextHop{
+			Hop: RTNextHop{},
+		}
+		copy(
+			(*(*[sizeOfRTNextHop]byte)(unsafe.Pointer(&nh.Hop)))[:],
+			(*(*[sizeOfRTNextHop]byte)(unsafe.Pointer(&mpData)))[i:],
```
Comment (mdlayher) [inline route.go:384]: Pretty sure this logic is incorrect. You'd want to point directly at the underlying array, but you're currently accessing the slice header itself.

I imagine it'd be something like `(*(*[sizeOfRTNextHop]byte)(unsafe.Pointer(&mpData[i:i+sizeOfRTNextHop])))[:]` instead.

This code is definitely going to need some fuzzing and careful tests.

File: route.go [hunk 2]
```diff
@@ -329,3 +335,101 @@ func (rm *RouteMetrics) encode(ae *netlink.AttributeEncoder) error {
...
+		// check again for a truncated message
+		if int(nh.Hop.Length) > mpPayloadSize {
+			return errInvalidRouteMessageAttr
+		}
+
+		// grab a new attributedecoder for the nested attributes
+		payloadStart := (i + sizeOfRTNextHop)
+		payloadEnd := (i + int(nh.Hop.Length))
+		nhDecoder, err := netlink.NewAttributeDecoder(mpData[payloadStart:payloadEnd])
+		nhDecoder.ByteOrder = nativeEndian
```
Comment (mdlayher) [inline route.go:396]: This is the default setting, you can remove this line.

File: route.go [hunk 3]
```diff
@@ -329,3 +335,101 @@ func (rm *RouteMetrics) encode(ae *netlink.AttributeEncoder) error {
...
+type NextHop struct {
+	Hop     RTNextHop // a rtnexthop struct
+	Gateway net.IP    // that struct's nested Gateway attribute
+	Value   []byte    // raw multipath `value` bytes (temporary for debugging)
+}
+
+// The RT_MULTIPATH netlink attribute contains a payload of `array of struct rtnexthop`
+type RTMultiPath []NextHop
+
+// consider instead creating an rtnetlink.MultiPathDecoder type
```
Comment (mdlayher) [inline route.go:357]: Seems okay as is to me.

**Review by mdlayher — COMMENTED** [review]
Looks like a good first pass, but the array walking logic will need careful testing and fuzzing to ensure there are no out of bounds panics and such.

File: route.go (hunk 1 above)
Comment (jsimonetti) [reply route.go:384]: Is the sizeOfRTNextHop value constant on every platform or is it possible that it varies across kernels/architectures? If so, we should use x/sys/unix for that constant.

Comment (djosephsen) [reply route.go:384]: the above syntax fails to build with `cannot take the address of mpData[i:i + sizeOfRTNextHop]`
I can get it building with an additional copy() by creating a next-hop-sized array to copy from eg.. 

```
var mpSlice [sizeOfRTNextHop]byte
        copy(mpSlice[:], mpData[i:i+sizeOfRTNextHop])

        copy(
            (*(*[sizeOfRTNextHop]byte)(unsafe.Pointer(&nh.Hop)))[:],
            (*(*[sizeOfRTNextHop]byte)(unsafe.Pointer(&mpSlice)))[:],
        )
```
is that sane? 

also, thanks, I'll change `sizeOfRTNextHop` to use the x/sys/unix constant.

File: route.go [hunk 4]
```diff
@@ -170,15 +170,17 @@ func (r *RouteService) List() ([]RouteMessage, error) {
...
-	OutIface uint32
-	Priority uint32
-	Table    uint32
-	Mark     uint32
-	Expires  *uint32
-	Metrics  *RouteMetrics
+	Dst       net.IP
+	Src       net.IP
+	Gateway   net.IP
+	MultPath  RTNextHop
```
Comment (jsimonetti) [inline route.go:176]: Typo I asume?

Comment (djosephsen) [reply route.go]: yeppers

**Comment by mdlayher** [comment]
Superseded by #88 .

**Pull request closed without merging**